feat(RHTAPWATCH-1204): use custom cert in push-snapshot

yftacherzog opened the pull request:
Adjusting the push-snapshot task so it can push to a repository that uses a self-signed certificate. It does that by mounting a config map containing custom CA certificate/bundle to a path which the OS uses for loading certificates.

yftacherzog commented:
/retest-required

mmalina commented on tasks/push-snapshot/README.md:
Please update `app.kubernetes.io/version` in the yaml file and add a change log entry.

yftacherzog replied:
@mmalina I missed this comment. Added it now.

mmalina replied:
Thanks. But can you use 6.3.0 instead, please? We use .z for bugfixes only and this is more of a new feature (but not a breaking one).

yftacherzog replied:
done

mmalina replied:
Thanks.

mmalina commented on tasks/push-snapshot/README.md:
```diff
@@ -4,12 +4,14 @@ Tekton task to push snapshot images to an image registry using `cosign copy`.
+| dataPath             | Path to the JSON string of the merged data to use in the data workspace   | No       |                      |
+| resultsDirPath       | Path to results directory in the data workspace                           | No       |                      |
+| retries              | Retry copy N times                                                        | Yes      | 0                    |
+| caTrustConfigMapName | The name of the ConfigMap to read CA bundle data from                     | Yes      | trusted-ca           |
```
Wouldn't it be better to use `""` as the default here? Assuming it would work with that.

yftacherzog replied:
If we do that, the user will have to provide value to that parameter, otherwise the task will not start, as it will be missing a name for the configmap associated with the volume.

Making the volume optional makes it possible for the default configmap not to exist.

mmalina replied:
OK, you're right, I checked and it won't work with an empty string. So I guess it should be ok like this.

mmalina commented on tasks/push-snapshot/push-snapshot.yaml:
```diff
@@ -179,3 +187,16 @@ spec:
         done
         echo -n "${RESULTS_JSON}" | tee $RESULTS_FILE
         printf 'Completed "%s" for "%s"\n\n' "$(context.task.name)" "$application"
+      volumeMounts:
```
Are you sure this will do the trick? 

This is what we do in our [Dockerfile](https://github.com/konflux-ci/release-service-utils/blob/main/Dockerfile#L51) to enable a new cert:
```
ADD data/certs/2015-IT-Root-CA.pem data/certs/2022-IT-Root-CA.pem /etc/pki/ca-trust/source/anchors/
RUN update-ca-trust
```
We use a different subdirectory, but also run `update-ca-trust`. So I wonder if what you propose will work.

yftacherzog replied:
I confirmed it to be working.

Doing it this way avoids the need to modify the task itself in order to copy things around and avoids writing into the container filesystem after the container is started.

mmalina replied:
OK, thanks for confirming. Can you add a test for this? You can create a custom-named configMap in tasks/push-snapshot/tests/pre-apply-task-hook.sh - similar what we do e.g. for create-pyxis-image:
```sh
# Create a dummy pyxis secret (and delete it first if it exists)
kubectl delete secret test-create-pyxis-image-cert --ignore-not-found
kubectl create secret generic test-create-pyxis-image-cert --from-literal=cert=mycert --from-literal=key=mykey
```
 And then use it in a test and then in tasks/push-snapshot/tests/mocks.sh in the cosign function you can add a special case for this where if some custom pull spec is used, you will confirm that the correct thing is mounted (based on the configMap you create and use in the test).

yftacherzog replied:
I'll give it a try. Thanks for the directions!

yftacherzog replied:
@mmalina have a look and let me know if this is what you had in mind.

mmalina replied:
That's exactly what I had in mind, great!

johnbieren commented on tasks/push-snapshot/tests/test-push-snapshot-mount-certs.yaml:
```diff
@@ -0,0 +1,125 @@
+apiVersion: tekton.dev/v1
+kind: Pipeline
+metadata:
+  name: test-push-snapshot
```
This should be `test-push-snapshot-mounted-certs`

yftacherzog replied:
done

johnbieren replied:
Sorry... you did what I said but I said the wrong thing :sweat_smile: 
It should be `test-push-snapshot-mount-certs`. The point I am getting at is the test names should be the same as the file name without the `.yaml` extension. The CI would actually timeout if this wasn't the case before

yftacherzog replied:
done

johnbieren commented:
Can you gpg sign the commit? https://github.com/konflux-ci/release-service/blob/main/CONTRIBUTING.md#signing-commits

yftacherzog committed:
feat(RHTAPWATCH-1204): use custom cert in push-snapshot

Adjusting the push-snapshot task so it can push to a repository that
uses a self-signed certificate. It does that by mounting a config map
containing custom CA certificate/bundle to a path which the OS uses
for loading certificates.

Signed-off-by: Yftach Herzog <yherzog@redhat.com>

yftacherzog commented:
> Can you gpg sign the commit? https://github.com/konflux-ci/release-service/blob/main/CONTRIBUTING.md#signing-commits

@johnbieren done. I think...

johnbieren commented:
Needs rebase then I think we can merge

yftacherzog commented:
> Needs rebase then I think we can merge

done
